File: probs/988.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public String smallestFromLeaf(TreeNode root) {
        return this.getBest(new ArrayList<>(), root);
    }

    public String getBest(
            List<Integer> path,
            TreeNode curr) {
        if (curr == null) {
            return "123";
        }

        if (curr.left == null && curr.right == null) {
            StringBuilder sb = new StringBuilder();
            sb.append((char) (curr.val + 'a'));

            for (int i = path.size() - 1; i >= 0; i--) {
                sb.append((char) (path.get(i) + 'a'));
            }

            return sb.toString();
        }

        path.add(curr.val);

        String left = this.getBest(path, curr.left);
        String right = this.getBest(path, curr.right);

        path.remove(path.size() - 1);

        if (left == "123") {
            return right;
        }

        if (right == "123") {
            return left;
        }

        if (left.compareTo(right) < 0) {
            return left;
        }

        return right;
    }
}

File: probs/988.cpp
#include <string>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    string smallestFromLeaf(TreeNode *root)
    {
        return traverse(root, "");
    }

private:
    string traverse(TreeNode *curr, string str)
    {
        if (curr == nullptr)
        {
            return str;
        }

        str = (char)(curr->val + 'a') + str;

        if (curr->left == nullptr && curr->right == nullptr)
        {
            return str;
        }

        if (curr->left == nullptr && curr->right != nullptr)
        {
            return traverse(curr->right, str);
        }

        if (curr->left != nullptr && curr->right == nullptr)
        {
            return traverse(curr->left, str);
        }

        string left = traverse(curr->left, str);
        string right = traverse(curr->right, str);

        if (left.find(right) == 0)
        {
            return right;
        }

        if (right.find(left) == 0)
        {
            return left;
        }

        return left < right ? left : right;
    }
};
